core-bundle/src/Doctrine/Backup/BackupManager.php

```diff
@@ -0,0 +1,311 @@
+    }
+
+    public function createCreateConfig(): CreateConfig
+    {
+        $config = new CreateConfig(Backup::createNewAtPath($this->backupDir));
+
+        return $config->withTablesToIgnore($this->tablesToIgnore);
+    }
+
+    /**
+     * @throws BackupManagerException
+     */
+    public function createRestoreConfig(): RestoreConfig
+    {
+        $latestBackup = $this->getLatestBackup();
+
+        if (null === $latestBackup) {
+            throw new BackupManagerException('No backups found.');
+        }
+
+        $config = new RestoreConfig($latestBackup);
+
+        return $config->withTablesToIgnore($this->tablesToIgnore);
+    }
+
+    public function getLatestBackup(): ?Backup
+    {
+        return $this->listBackups()[0] ?? null;
+    }
+
+    /**
+     * Lists all backups (latest one first).
+     *
+     * @return array<Backup>
+     */
+    public function listBackups(): array
+    {
+        (new Filesystem())->mkdir($this->backupDir); // Ensures, the directory exists
+
+        $backups = [];
+        $files = Finder::create()
+            ->files() // TODO: depth == 0?
+            ->in($this->backupDir)
+            ->ignoreDotFiles(true)
+            ->ignoreVCS(true)
+        ;
+
+        foreach ($files as $file) {
+            $backups[] = new Backup($file->getPathname());
+        }
+
+        uasort($backups, static fn (Backup $a, Backup $b) => $b->getCreatedAt() <=> $a->getCreatedAt());
+
+        return $backups;
+    }
+
+    /**
+     * @throws BackupManagerException
+     */
+    public function create(CreateConfig $config): void
+    {
+        $this->executeTransactional(
+            function () use ($config): void {
+                $this->doCreate($config);
+            }
+        );
+    }
+
+    /**
+     * @throws BackupManagerException
+     */
+    public function restore(RestoreConfig $config): void
+    {
+        $this->executeTransactional(
+            function () use ($config): void {
+                $this->doRestore($config);
+            }
+        );
+    }
+
+    /**
+     * @throws BackupManagerException
+     */
+    private function executeTransactional(\Closure $func): void
+    {
+        $isAutoCommit = $this->connection->isAutoCommit();
+
+        if ($isAutoCommit) {
+            $this->connection->setAutoCommit(false);
+        }
+
+        try {
+            $this->connection->transactional($func);
+        } catch (BackupManagerException $e) {
+            throw $e;
+        } catch (\Throwable $t) {
+            throw new BackupManagerException($t->getMessage(), 0, $t);
+        }
+
+        if ($isAutoCommit) {
+            $this->connection->setAutoCommit(true);
+        }
+    }
+
+    /**
+     * @throws BackupManagerException
+     */
+    private function doCreate(CreateConfig $config): void
+    {
+        $backup = $config->getBackup();
+
+        // Ensure the target file exists and is empty
+        (new Filesystem())->dumpFile($backup->getFilepath(), '');
```

Comment:
I think we shouldn't overwrite if the file already exists 🤔
also, `fopen` already does that and truncates the file?

Reply:
Yes we should, that's exactly why I'm doing this here. And I'm using `Filesystem` because it's safer :)